lab-will/app/entry.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+require('./scss/reset.scss');
+require('./scss/main.scss');
+
+const angular = require('angular');
+const cowsay = require('cowsay-browser');
+
+// Create the entire app
+const cowsayApp = angular.module('cowsayApp', []);
+
+// Create and name the CowsayController
+cowsayApp.controller('CowsayController', ['$log', CowsayController]);
+
+// Define the CowsayController constructor
+function CowsayController($log) {
+  // Debug
+  $log.debug('CowsayController');
+
+  // Properties
+  this.title = 'Welcome to Cowville!';
+  this.history = [];
+
+  // History Stack
+  cowsay.list( (err, cowfiles) => {
+    this.cowfiles = cowfiles;
+    this.current = cowfiles[0];
+  });
+
+  // Update message
+  this.update = function(input) {
+    $log.debug('cowsayCtrl.update()');
+    return cowsay.say({ text: input || 'moooooooo', f: this.current });
+  };
+
+  // Make the cow say the thing
+  this.speak = function(input) {
+    $log.debug('cowsayCtrl.speak()');
+
+    // Update message
+    this.spoken = this.update(input);
+
+    // Add this message to the history stack
+    this.history.push(this.spoken);
+  };
+
+  // Undo last message
+  this.undo = function() {
+    $log.debug('cowsayCtrl.undo()');
+
+    // Remove from history
+    this.history.pop();
+    this.spoken = this.history.pop() || '';
+  };
+};
```

Comment:
Just a little clarification.  History is a stack containing all of the cows we have built.  The top cow on the stack is the current one being displayed.  When we click undo we want to see the previous cow.  So the first pop() pops off the current cow.  The second pop() pops off the previous cow and sets it equal to this.spoken, which is the cow we are displaying.